crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.num_qubits()
+}
+
+/// @ingroup QkSparseObservable
+/// Get the number of bit terms/indices in the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return The number of terms in the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint64_t len = qk_obs_len(obs);  // len==0, as there are no non-trivial bit terms
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_len(obs: *const SparseObservable) -> u64 {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.bit_terms().len() as u64
+}
+
+/// @ingroup QkSparseObservable
+/// Get a pointer to the coefficients.
+///
+/// This can be used to read and modify the observable's coefficients.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to the coefficients.
+///
+/// # Example
+///     
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint64_t num_terms = qk_obs_num_terms(obs);
+///     complex double *coeffs = qk_obs_coeffs(obs);
+///
+///     for (uint64_t i = 0; i < num_terms; i++) {
+///         printf("%f + i%f\n", creal(coeffs[i]), cimag(coeffs[i]));
+///     }
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_coeffs(obs: *mut SparseObservable) -> *mut Complex64 {
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+    &mut obs.coeffs_mut()[0]
+}
+
+/// @ingroup QkSparseObservable
+/// Get a pointer to the indices.
+///
+/// This can be used to read and modify the observable's indices.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to the indices.
+///
+/// # Example
+///
+///     uint32_t num_qubits = 100;
+///     QkSparseObservable *obs = qk_obs_zero(num_qubits);
+///
+///     complex double coeff = 1;
+///     QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+///     uint32_t indices[3] = {0, 1, 2};
+///     QkSparseTerm term = {&coeff, 3, bit_terms, indices, num_qubits};
+///     qk_obs_add_term(obs, &term);
+///
+///     uint64_t len = qk_obs_len(obs);
+///     uint32_t *indices = qk_obs_indices(obs);
+///
+///     for (uint64_t i = 0; i < len; i++) {
+///         printf("index %i: %i\n", i, indices[i]);
+///     }
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_indices(obs: *mut SparseObservable) -> *mut u32 {
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+
+    // this is unsafe as we can no longer ensure the indices are within
+    // range of the observable's number of qubits
+    unsafe { &mut obs.indices_mut()[0] }
```

Comment:
Since `obs` becomes a `&mut SparseObservable`, we've already assumed that the `indices` field as-pointed-to is valid, so the unsafe-ness shouldn't be about taking the index, it should only be about retrieving a mutable reference to `indices`.  Accessing the zeroth element of that will be bounds checked by Rust, but also, we should just be able to do `unsafe { obs.indices_mut() }.as_mut_ptr()`.

(Comment repeats elsewhere in similar methods.)

Reply:
Updated in bc376c6